### PR TITLE
Up2

### PR DESCRIPTION
**Pull request by lyakh**
This is a re-spin of #1240 with minimum changes, the largest being a switch to the common HDMI codec drover.

### CODE REVIEW

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 1]
```diff
@@ -0,0 +1,474 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright(c) 2018-2019 Intel Corporation.
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_pcm512x.c]: 2018-2020

File: sound/soc/intel/boards/Kconfig [hunk 2]
```diff
@@ -483,6 +471,20 @@ config SND_SOC_INTEL_SOF_RT5682_MACH
 	   with rt5682 codec.
 	   Say Y if you have such a device.
 	   If unsure select "N".
+
+config SND_SOC_INTEL_SOF_PCM512x_MACH
+	tristate "Broxton with TI PCM512x codec"
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: SOF with TI PCM......

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+
+static struct snd_soc_dai_link_component platform_component[] = {
+	{
+		/* name might be overridden during probe */
+		.name = "0000:00:1f.3"
+	}
+};
+
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI) || \
+	IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_pcm512x.c]: do we support HDAC_HDMI? It's not selected in the Kconfig?

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+	hdmi_num = 0;
+	if (soc_intel_is_byt() || soc_intel_is_cht()) {
+		is_legacy_cpu = true;
+		dmic_be_num = 0;
+		/* default quirk for legacy cpu */
+		sof_pcm512x_quirk = SOF_PCM512X_SSP_CODEC(2);
+	} else {
+		dmic_be_num = 2;
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI) || \
+	IS_ENABLED(CONFIG_SND_SOC_HDAC_HDMI)
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_pcm512x.c]: Do we support HDAC_HDMI?

File: sound/soc/intel/common/soc-acpi-intel-cht-match.c [hunk 5]
```diff
@@ -174,6 +174,13 @@ struct snd_soc_acpi_mach  snd_soc_acpi_intel_cherrytrail_machines[] = {
 		.sof_fw_filename = "sof-cht.ri",
 		.sof_tplg_filename = "sof-cht-cx2072x.tplg",
 	},
+	{
+		.id = "104C5122",
+		.drv_name = "sof_pcm512x",
+		.sof_fw_filename = "sof-cht.ri",
+		.sof_tplg_filename = "sof-cht-src-50khz-pcm512x.tplg",
+	},
+
```
Comment (plbossart) [inline sound/soc/intel/common/soc-acpi-intel-cht-match.c:183]: can you split this in another patch, that would help me submit upstream? Thanks!

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_pcm512x.c]: @plbossart @lyakh Ack, let's remove hdac-hdmi support from this machien driver. We have no reason to keep it. So CONFIG_SND_HDA_CODEC_HDMI only.

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+	int i = 0;
+
+	/* HDMI is not supported by SOF on Baytrail/CherryTrail */
+	if (is_legacy_cpu)
+		return 0;
+
+	if (list_empty(&ctx->hdmi_pcm_list))
+		return -EINVAL;
+
+	if (ctx->common_hdmi_codec_drv) {
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_pcm512x.c]: To future-proof, we should handle  (!ctx->common_hdmi_codec_drv)  case also such that we "return 0" i.e. HDMI is not supported rather than reporting an error. Applies to the hdmi_pcm_list check as well.

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+		pcm = list_first_entry(&ctx->hdmi_pcm_list, struct sof_hdmi_pcm,
+				       head);
+		component = pcm->codec_dai->component;
+		return hda_dsp_hdmi_build_controls(card, component);
+	}
+
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_pcm512x.c]: So below this line is only for hdac-hdmi and can be dropped.

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 4 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_pcm512x.c]: Ack, here as well, drop hdac-hdmi.

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 6 above)
Comment (lyakh) [reply sound/soc/intel/boards/sof_pcm512x.c]: @kv2019i do we need to check `mach->mach_params.common_hdmi_codec_drv` at all? I thought if we drop HDAC_HDMI, we ignore that flag completely? Or should we check and if it *isn't* set, skip HDMI registration and return 0?

File: sound/soc/intel/common/soc-acpi-intel-cht-match.c (hunk 5 above)
Comment (lyakh) [reply sound/soc/intel/common/soc-acpi-intel-cht-match.c:183]: @plbossart you mean split patch 2 of the series into 2 parts - separately for soc-acpi-intel-bxt-match.c and soc-acpi-intel-cht-match.c? Ok, will do

Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-cht-match.c:183]: yes, thank you

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+	if (ret)
+		dev_err(card->dev, "DMic map addition failed: %d\n", ret);
+
+	return ret;
+}
+
+/* sof audio machine driver for pcm512x codec */
+static struct snd_soc_card sof_audio_card_pcm512x = {
+	.name = "sof_pcm512x",
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_pcm512x.c]: @lyakh we shoudl remove the "sof_" prefix here. The "sof-" will be added by the core

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 6 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_pcm512x.c]: @lyakh We still need to check it as long as we support it for some platforms. Basicly only case it would be false is a kernel configuration error (machine driver should bring in right modules via kconfig), but it's still needs to be checked.

File: sound/soc/intel/common/soc-acpi-intel-cht-match.c [hunk 9]
```diff
@@ -174,6 +174,13 @@ struct snd_soc_acpi_mach  snd_soc_acpi_intel_cherrytrail_machines[] = {
 		.sof_fw_filename = "sof-cht.ri",
 		.sof_tplg_filename = "sof-cht-cx2072x.tplg",
 	},
+	{
+		.id = "104C5122",
+		.drv_name = "sof_pcm512x",
```
Comment (ranj063) [inline sound/soc/intel/common/soc-acpi-intel-cht-match.c:179]: @plbossart with the introduction of default_fw_filename in the machine desc in SOF, we dont actually need the sof_fw_filename field in the acpi_mach anymore. Should we consider removing it?

Comment (ranj063) [reply sound/soc/intel/common/soc-acpi-intel-cht-match.c:179]: Oh btw Im asking for the future, it doesnt affect this PR.

**Comment by lyakh** [comment]
> sign

@ranj063 thanks for the review! Sorry, don't quite understand, what's wrong with signoffs?

**Comment by ranj063** [comment]
> > sign
> 
> @ranj063 thanks for the review! Sorry, don't quite understand, what's wrong with signoffs?

"Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>
[guennadi.liakhovetski@linux.intel.com: switch to common HDMI]
Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>"

This is what I see. Whats the "[guennadi.liakhovetski@linux.intel.com: switch to common HDMI]" in between the 2 sign-off's?

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 8 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_pcm512x.c]: yes, nice catch

File: sound/soc/intel/common/soc-acpi-intel-cht-match.c (hunk 9 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-cht-match.c:179]: @ranj063 I don't mind if we use the default firmware if the machine driver does not provide a firmware file name, but I'd really like to keep the ability to use a different firmware for a specific platform, e.g. to allow for experiments before everything is merged.
in other works, ok to use a default but not ok to only use a default.

**Comment by lyakh** [comment]
> > > sign
> > 
> > 
> > @ranj063 thanks for the review! Sorry, don't quite understand, what's wrong with signoffs?
> 
> "Signed-off-by: Pierre-Louis Bossart [pierre-louis.bossart@linux.intel.com](mailto:pierre-louis.bossart@linux.intel.com)
> [[guennadi.liakhovetski@linux.intel.com](mailto:guennadi.liakhovetski@linux.intel.com): switch to common HDMI]
> Signed-off-by: Guennadi Liakhovetski [guennadi.liakhovetski@linux.intel.com](mailto:guennadi.liakhovetski@linux.intel.com)"
> 
> This is what I see. Whats the "[[guennadi.liakhovetski@linux.intel.com](mailto:guennadi.liakhovetski@linux.intel.com): switch to common HDMI]" in between the 2 sign-off's?

Yes, that's how it should be. That means I've got that patch from @plbossart with his SoB, made some changes, and now I'm submitting it with my SoB.

File: sound/soc/intel/common/soc-acpi-intel-cht-match.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/intel/common/soc-acpi-intel-cht-match.c:179]: @plbossart OK makes sense then in that case I guess we do need a minor change in the SOF driver to use the filename provided in the ACPI match table if provided to make use of it. I'll send a PR

**Comment by plbossart** [comment]
> This is what I see. Whats the "[[guennadi.liakhovetski@linux.intel.com](mailto:guennadi.liakhovetski@linux.intel.com): switch to common HDMI]" in between the 2 sign-off's?

Guennadi has a very strong sense of intellectual honesty and I guess wanted to clarify his contribution vs. mine. It's fine for me to remove this mention, as well as the initial part mentioning the sof-rt5682 lineage.

**Comment by ranj063** [comment]
> > This is what I see. Whats the "[[guennadi.liakhovetski@linux.intel.com](mailto:guennadi.liakhovetski@linux.intel.com): switch to common HDMI]" in between the 2 sign-off's?
> 
> Guennadi has a very strong sense of intellectual honesty and I guess wanted to clarify his contribution vs. mine. It's fine for me to remove this mention, as well as the initial part mentioning the sof-rt5682 lineage.

ha nice tip to learn for today. I should keep this in mind for the future! BTW is there a way to add this mention automatically?

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_pcm512x.c]: Not following @kv2019i, we don't select HDAC_HDMI here and don't make references to hdac_hdmi provided symbols, when could this be problematic (I am only talking about this specific machine driver)

**Comment by plbossart** [comment]
@lyakh if you can provide an update by my morning it'd be nice, I'll group this with the rest of the machine driver updates in one upstream contribution. The code looks good and there's only minor things to fix.

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 6 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_pcm512x.c]: @plbossart The user could forcibly disable common_hdmi_codec mode (via kernel module parameter). It's an invalid config for sure, but we should not at least crash in this case. hda_dsp_hdmi_build_controls() expects the codec driver to be loaded, which it wouldn't be in this case.

Comment (plbossart) [reply sound/soc/intel/boards/sof_pcm512x.c]: @kv2019i ok, I forgot about the kernel parameter. Do we actually still have a reason to keep it? It generates more problems than solutions I am afraid.

Comment (kv2019i) [reply sound/soc/intel/boards/sof_pcm512x.c]: @plbossart We still have some issues with the common driver on Chrome devices, so I'm afraid we have to keep it at least for a few stable cycles. I'm afraid. But yeah, supporting both has generated a lot of work. If/when we get the new UCM2 to longterm distro releases, doing this kind of changes will be easier in the future.

Comment (plbossart) [reply sound/soc/intel/boards/sof_pcm512x.c]: ok, but this specific driver is mostly for tests, so I suggest we don't bother with the 'uncommmon' hdmi driver.

Comment (lyakh) [reply sound/soc/intel/boards/sof_pcm512x.c]: @plbossart I'm trying to prepare a PR for you, so I understand I'll remove `mach->mach_params.common_hdmi_codec_drv` completely

**Comment by lyakh** [comment]
> @lyakh if you can provide an update by my morning it'd be nice, I'll group this with the rest of the machine driver updates in one upstream contribution. The code looks good and there's only minor things to fix.

@plbossart let me know, if there's still anything to improve. The CI failures are 3 silly checkpatch "checks" and runtime failures - clearly on Up2 because the topology has to be updated, but unclear why on other platforms too.

**Comment by lyakh** [comment]
A short status update: I tested DMIC and line-in capture with a good result. Previously I was getting corrupted and channel-unbalanced audio from line-in. Now both sources record audio correctly. Although using "arecord -d N" without explicitly specifying the sampling rate causes the capture to be about 6 times shorter, but still with correct audio. Specifying a 48kHz sampling rate fixes the capture duration. Also the line-in captured audio is very quiet. The respective topology changes are in https://github.com/thesofproject/sof/pull/2301

**Comment by plbossart** [comment]
> A short status update: I tested DMIC and line-in capture with a good result. Previously I was getting corrupted and channel-unbalanced audio from line-in. Now both sources record audio correctly. Although using "arecord -d N" without explicitly specifying the sampling rate causes the capture to be about 6 times shorter, but still with correct audio. Specifying a 48kHz sampling rate fixes the capture duration. Also the line-in captured audio is very quiet. The respective topology changes are in [thesofproject/sof#2301](https://github.com/thesofproject/sof/pull/2301)

Thanks! we probably have a missing constraint in the hw_params and end-up recording at 8kHz but providing 48kHz data. Could it be that the topology says we can use whatever rate we want, but that's not quite right here?

**Comment by lyakh** [comment]
> > A short status update: I tested DMIC and line-in capture with a good result. Previously I was getting corrupted and channel-unbalanced audio from line-in. Now both sources record audio correctly. Although using "arecord -d N" without explicitly specifying the sampling rate causes the capture to be about 6 times shorter, but still with correct audio. Specifying a 48kHz sampling rate fixes the capture duration. Also the line-in captured audio is very quiet. The respective topology changes are in [thesofproject/sof#2301](https://github.com/thesofproject/sof/pull/2301)
> 
> Thanks! we probably have a missing constraint in the hw_params and end-up recording at 8kHz but providing 48kHz data. Could it be that the topology says we can use whatever rate we want, but that's not quite right here?

@plbossart arecord is complaining:
```
Warning: rate is not accurate (requested = 8000Hz, got = 48000Hz)
```

**Comment by plbossart** [comment]
> > > A short status update: I tested DMIC and line-in capture with a good result. Previously I was getting corrupted and channel-unbalanced audio from line-in. Now both sources record audio correctly. Although using "arecord -d N" without explicitly specifying the sampling rate causes the capture to be about 6 times shorter, but still with correct audio. Specifying a 48kHz sampling rate fixes the capture duration. Also the line-in captured audio is very quiet. The respective topology changes are in [thesofproject/sof#2301](https://github.com/thesofproject/sof/pull/2301)
> > 
> > 
> > Thanks! we probably have a missing constraint in the hw_params and end-up recording at 8kHz but providing 48kHz data. Could it be that the topology says we can use whatever rate we want, but that's not quite right here?
> 
> @plbossart arecord is complaining:
> 
> ```
> Warning: rate is not accurate (requested = 8000Hz, got = 48000Hz)
> ```

right, but we should return an error here. Something's missing. @ranj063 any ideas?

**Comment by ranj063** [comment]
> right, but we should return an error here. Something's missing. @ranj063 any ideas?

@plbossart the topology limits the rate min/max to 48000kHz in our case but I think we set the rate min/max only during the be_hw_params_fixup. I'm not sure about how to throw an error in this case. @singalsu ?

File: sound/soc/intel/common/soc-acpi-intel-cht-match.c (hunk 9 above)
Comment (ranj063) [reply sound/soc/intel/common/soc-acpi-intel-cht-match.c:179]: @plbossart I just realised something while working on the multi-client implementation. With multiple clients with separate machine drivers, does it make sense to have the fw_filename specified in each of them individually? Also, we will be probing the SOF DSP device first before we even probe the machine driver and the FW will be loaded when the SOF device is probed.

Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-cht-match.c:179]: well we need to have a single firmware, but each client might have a subset of the topologies.

Comment (ranj063) [reply sound/soc/intel/common/soc-acpi-intel-cht-match.c:179]: OK thats what I hae now. So then the only fw file that will be used is the default_fw_filename in the sof dev desc.

**Comment by singalsu** [comment]
> > right, but we should return an error here. Something's missing. @ranj063 any ideas?
> 
> @plbossart the topology limits the rate min/max to 48000kHz in our case but I think we set the rate min/max only during the be_hw_params_fixup. I'm not sure about how to throw an error in this case. @singalsu ?

Sorry, which audio device and which topology was used? Is it this https://github.com/thesofproject/sof/pull/2301 ?

**Comment by lyakh** [comment]
> > > right, but we should return an error here. Something's missing. @ranj063 any ideas?
> > 
> > 
> > @plbossart the topology limits the rate min/max to 48000kHz in our case but I think we set the rate min/max only during the be_hw_params_fixup. I'm not sure about how to throw an error in this case. @singalsu ?
> 
> Sorry, which audio device and which topology was used? Is it this [thesofproject/sof#2301](https://github.com/thesofproject/sof/pull/2301) ?

@singalsu yes, that one

**Comment by singalsu** [comment]
@kv2019i Should we instead of triggering warning from arecord like "Warning: rate is not accurate (requested = 8000Hz, got = 48000Hz)" stop into error? Is there a common practice in ALSA for such?

**Comment by paulstelian97** [comment]
@singalsu I have noticed that when I play a file through plughw this is compensated and I can play 44100Hz files without any problems on my 48000Hz-only topology. So as long as you report the actual rate correctly it's fine even though you may lose power savings. Sure it's probably best to support more rates but as an alternative (i.MX is currently on the alternative) this can work.

**Comment by kv2019i** [comment]
@singalsu wrote:
> @kv2019i Should we instead of triggering warning from arecord like "Warning: rate is not 
> accurate (requested = 8000Hz, got = 48000Hz)" stop into error? Is there a common practice in 
> ALSA for such?

This is expected behaviour. A hw driver can only expose the sampling rates it actually supports. There are many, many devices that only support a single sampling rate, so this is not very unusual eiter. Application is using snd_pcm_hw_params_set_rate_near() and correctly sets the result. Aplay is a devel application, so it decides to only emit a warning. 

If one is using aplay/arecord for "production use", the plugin layer should be used. It will apply resampling if needed. Most users will use an audio server like Pulseaudio and JACK which handle resampling themselves. Here user is using direct hw access and should know what she/he is doing. ;)

IOW, this is expected behaviour.

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/types.h>
+#include <sound/core.h>
+#include <sound/jack.h>
+#include <sound/pcm.h>
+#include <sound/pcm_params.h>
+#include <sound/soc.h>
+#include <sound/soc-acpi.h>
+#include "../../codecs/hdac_hdmi.h"
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_pcm512x.c]: You can also drop this "hdac_hdmi.h" now.

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+{
+	struct sof_card_private *ctx = snd_soc_card_get_drvdata(card);
+	struct sof_hdmi_pcm *pcm;
+
+	/* HDMI is not supported by SOF on Baytrail/CherryTrail */
+	if (is_legacy_cpu)
+		return 0;
+
+	if (list_empty(&ctx->hdmi_pcm_list))
+		return -EINVAL;
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_pcm512x.c:137]: Just a note that in generic HDA driver, I changed this so that empty hdmi list is not an error anymore. But this is ok for UP2, this is anyway machine driver specific policy.

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+static int aif1_startup(struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_component *codec = rtd->codec_dai->component;
+
+	snd_soc_component_update_bits(codec, PCM512x_GPIO_CONTROL_1,
+				      0x08, 0x08);
+
+	return snd_pcm_hw_constraint_single(substream->runtime,
+			SNDRV_PCM_HW_PARAM_RATE, 48000);
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_pcm512x.c]: Ah, @lyakh @singalsu , found it! It's of course this that is the problem. Here a single-rate constraint to SSP5 is put in place for the substream. This will constraint the selected rate for FE PCM and e.g. it's impossible to do
"aplay --dump-hw-params -r8000 -Dhw:0,4 8khz-test.wav". 
If you just "return 0" here and not set any constraint, playback works as expected and SRC is done on DSP. 
@plbossart Anything we break if we just leave this out?

File: sound/soc/intel/boards/Kconfig [hunk 13]
```diff
@@ -525,7 +527,6 @@ config SND_SOC_INTEL_TGL_RT711_RT1308_MACH
 	select SND_SOC_RT1308
 	select SND_SOC_DMIC
 	select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_AUDIO_CODEC
-	select SND_SOC_HDAC_HDMI if SND_SOC_SOF_HDA_LINK
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: this looks unrelated?

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_pcm512x.c]: we are still in 2020

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 12 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_pcm512x.c]: I am not sure why this is here indeed. The goal of this routine in only to turn on the LED.
Let's remove this.

I am not sure however if the SRC will be done on the low-latency path, it should only work with 48kHz?

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 14]
```diff
@@ -0,0 +1,439 @@
...
+		.dai_name = "pcm512x-hifi",
+	}
+};
+
+static struct snd_soc_dai_link_component dmic_component[] = {
+	{
+		.name = "dmic-codec",
+		.dai_name = "dmic-hifi",
+	}
+};
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_pcm512x.c]: can we use the 'modern style' components definition, as in all other machine drivers?

File: sound/soc/intel/boards/Kconfig [hunk 15]
```diff
@@ -483,6 +471,20 @@ config SND_SOC_INTEL_SOF_RT5682_MACH
 	   with rt5682 codec.
 	   Say Y if you have such a device.
 	   If unsure select "N".
+
+config SND_SOC_INTEL_SOF_PCM512x_MACH
+	tristate "SOF with TI PCM512x codec"
+	depends on I2C && ACPI
+	depends on (SND_SOC_SOF_HDA_LINK && (MFD_INTEL_LPSS || COMPILE_TEST)) ||\
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: @kv2019i this should be HDA_AUDIO_CODEC. no? we don't support the plain vanilla hdac_hdmi stuff so probably need to restrict more here.

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 10 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_pcm512x.c]: this too

File: sound/soc/intel/boards/Kconfig (hunk 15 above)
Comment (lyakh) [reply sound/soc/intel/boards/Kconfig]: @plbossart there's a
```
select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_AUDIO_CODEC
```
below, that's also what other drivers do. You're proposing
```
	depends on (SND_SOC_SOF_HDA_AUDIO_CODEC && (MFD_INTEL_LPSS || COMPILE_TEST)) ||\
		   (SND_SOC_SOF_BAYTRAIL && (X86_INTEL_LPSS || COMPILE_TEST))
	select SND_SOC_PCM512x_I2C
	select SND_HDA_CODEC_HDMI if SND_SOC_SOF_HDA_AUDIO_CODEC
```
?

Comment (plbossart) [reply sound/soc/intel/boards/Kconfig]: yes, the snippet above is what I meant. The conditional select if needed for the BAYTRAIL case.

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 12 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_pcm512x.c]: @plbossart wrote:
> I am not sure however if the SRC will be done on the low-latency path, it should only work with 48kHz?

It works correctly still even if the constraints is left out here. I.e. pcm0/low-latency only supports 48kHz, while pcm4/media support the full range.

File: sound/soc/intel/boards/Kconfig (hunk 15 above)
Comment (kv2019i) [reply sound/soc/intel/boards/Kconfig]: @plbossart @lyakh  Ack, this looks good.

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 14 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_pcm512x.c]: @lyakh I think this is the only open left -> using 'modern style' for components.

Comment (lyakh) [reply sound/soc/intel/boards/sof_pcm512x.c]: @kv2019i mmh, are you sure you're looking at the newest version? I updated two locations, but maybe I missed some more?

File: sound/soc/intel/boards/sof_pcm512x.c [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+
+	links[id].id = id;
+	links[id].codecs = pcm512x_component;
+	links[id].num_codecs = ARRAY_SIZE(pcm512x_component);
+	links[id].platforms = platform_component;
+	links[id].num_platforms = ARRAY_SIZE(platform_component);
+	links[id].init = sof_pcm512x_codec_init;
+	links[id].ops = &sof_pcm512x_ops;
+	links[id].nonatomic = true;
+	links[id].dpcm_playback = 1;
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_pcm512x.c:245]: @lyakh I understood Pierre's earlier comment about "modern style" is to replace this dynamic buildup of links with something more like done e.g. in bxt_rt298.c . I.e. use SND_SOC_DAILINK_REG() to populate a static array and then only have some conditional logic here to modify the entries as per specifis of the platforms.

But yeah, as this has more conditional logic (old legacy platforms, dmic, hdmi), so not sure how well it fits here. @plbossart any clarifications?

For my part, it would be good to use future-proof style when adding new drivers, but I'm honestly not sure how bad this "non-modern" style is for cases like this.
UPDATE: you already updated to use modern macros for the DEFS. Maybe this is indeed good enough.

File: sound/soc/intel/boards/sof_pcm512x.c (hunk 14 above)
Comment (kv2019i) [reply sound/soc/intel/boards/sof_pcm512x.c]: @lyakh Ack. Indeed, if this comment was only for the _DEFS and not the _REGs, then this is indeed already done.

**Comment by plbossart** [comment]
But now I am confused as to why CI fails? @lyakh do we have a cross-dependency with topology changes?

**Comment by kv2019i** [comment]
@plbossart wrote:
> But now I am confused as to why CI fails? @lyakh do we have a cross-dependency with topology changes?

Yes, the FW patch changes the DAI id numbering scheme and this Linux patch adjusts to the FW patch. So the two patches need to be merged at the same time.
